arlo_server/jurisdictions.py

```diff
@@ -1,35 +1,139 @@
+        .filter(AuditBoard.signed_off_at.isnot(None))
+        .group_by(AuditBoard.jurisdiction_id)
+        .values(AuditBoard.jurisdiction_id, func.count())
+    )
+
+    sampled_ballot_count_by_jurisdiction = dict(
+        SampledBallotDraw.query.filter_by(round_id=round.id)
+        .join(Batch)
+        .group_by(Batch.jurisdiction_id)
+        .values(Batch.jurisdiction_id, func.count())
+    )
+    audited_ballot_count_by_jurisdiction = dict(
+        SampledBallotDraw.query.filter_by(round_id=round.id)
+        .join(SampledBallot)
+        .filter(SampledBallot.vote.isnot(None))
+        .join(Batch)
+        .group_by(Batch.jurisdiction_id)
+        .values(Batch.jurisdiction_id, func.count())
+    )
+
+    def num_ballots_sampled(jurisdiction_id: str) -> int:
+        return sampled_ballot_count_by_jurisdiction.get(jurisdiction_id, 0)
+
+    # NOT_STARTED = the jurisdiction hasn’t set up any audit boards
+    # IN_PROGRESS = the audit boards are set up
+    # COMPLETE = all of the audit boards have signed off on their ballots
+    def status(jurisdiction_id: str) -> JurisdictionStatus:
+        num_set_up = audit_boards_set_up.get(jurisdiction_id, 0)
+        num_signed_off = audit_boards_signed_off.get(jurisdiction_id, 0)
+        num_sampled = num_ballots_sampled(jurisdiction_id)
+
+        # Special case: jurisdictions that don't get any ballots assigned are
+        # COMPLETE from the get-go
+        if num_sampled == 0:
+            return JurisdictionStatus.COMPLETE
+        elif num_set_up == 0:
+            return JurisdictionStatus.NOT_STARTED
+        elif num_signed_off != num_set_up:
+            return JurisdictionStatus.IN_PROGRESS
+        else:
+            return JurisdictionStatus.COMPLETE
+
+    def num_ballots_audited(jurisdiction_id: str) -> int:
+        if online:
+            return audited_ballot_count_by_jurisdiction.get(jurisdiction_id, 0)
+        else:
+            # For offline audits, we can't track incremental progress on
+            # auditing ballots, so we just report 0 ballots until the round is
+            # complete, then report all the sampled ballots.
+            return (
+                0
+                if status(jurisdiction_id) != JurisdictionStatus.COMPLETE
+                else num_ballots_sampled(jurisdiction_id)
+            )
+
+    return {
+        j.id: {
+            "status": status(j.id),
+            "numBallotsSampled": num_ballots_sampled(j.id),
+            "numBallotsAudited": num_ballots_audited(j.id),
+        }
+        for j in jurisdictions
     }
 
 
 @app.route("/election/<election_id>/jurisdiction", methods=["GET"])
 @with_election_access(UserType.AUDIT_ADMIN)
 def list_jurisdictions(election: Election):
-    jurisdictions = (
-        Jurisdiction.query.filter_by(election_id=election.id)
-        .order_by(Jurisdiction.name)
-        .all()
+    jurisdictions = sorted(election.jurisdictions, key=lambda j: j.name)
```

Comment:
could do this straight in the database query that has to run anyways:

https://docs.sqlalchemy.org/en/13/orm/extensions/declarative/relationships.html

Reply:
Love it! fa554f0